### PR TITLE
Refactor: read mode keeps sheaf state only

### PR DESCRIPTION
**Pull request by rtshkmr**
pls refer to https://github.com/ve1ld/vyasa/pull/125#pullrequestreview-2395293022 for better context 

### Frankly, just read the whole read context on emacs directly, much has been cleaned up 

<details>
<summary>my notes on this -- this is my default level of frictionless notes now</summary>

# org-export

1.  [update the current read mode, use sheaf<sub>ui</sub> to encapsulate the current ui structs](#orgfdfcd5f)
    1.  [Notes on some of the changes](#org2028ed5)
    2.  [Refactor Tasks](#orgeb93e45)
        1.  [the reflector mutator doesn&rsquo;t seem to be working](#org5220093)
        2.  [the nested form problem](#orgb196120)
        3.  [Granular UI State management](#org1f2640d)
        4.  [ui function interface](#orgd6e6faa)

<a id="orgfdfcd5f"></a>

# DONE update the current read mode, use sheaf<sub>ui</sub> to encapsulate the current ui structs

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">
<caption class="t-above"><span class="table-number">Table 1:</span> Clock summary at <span class="timestamp-wrapper"><span class="timestamp">[2024-10-25 Fri 22:34]</span></span></caption>

<colgroup>
<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />

<col  class="org-right" />

<col  class="org-right" />

<col  class="org-left" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Headline</th>
<th scope="col" class="org-left">Time</th>
<th scope="col" class="org-left">&#xa0;</th>
<th scope="col" class="org-left">&#xa0;</th>
<th scope="col" class="org-left">&#xa0;</th>
<th scope="col" class="org-right">&#xa0;</th>
<th scope="col" class="org-right">&#xa0;</th>
<th scope="col" class="org-left">&#xa0;</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left"><b>Total time</b></td>
<td class="org-left"><b>17:46</b></td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>
</tbody>
<tbody>
<tr>
<td class="org-left">&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;update the current read mode, use&#x2026;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-right">17:46</td>
<td class="org-right">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;Refactor Tasks</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-right">17:46</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;the reflector mutator doesn&rsquo;t seem to&#x2026;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-left">9:14</td>
</tr>

<tr>
<td class="org-left">&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;the nested form problem</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-left">3:33</td>
</tr>

<tr>
<td class="org-left">&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;Granular UI State management</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-right">&#xa0;</td>
<td class="org-left">4:59</td>
</tr>
</tbody>
</table>

-   CLOSING NOTE <span class="timestamp-wrapper"><span class="timestamp">[2024-10-25 Fri 22:31] </span></span>   
    This is done now

This will allow the sheaf ui structs to support what is needed, that API can be enriched

<a id="org2028ed5"></a>

## Notes on some of the changes

1.  [BUGFIX]: traits updated in the changeset so that the published state when publishing sheaf actually shows up on the db write
2.  the event names have been updated with appropriate prefixes:
    -   <struct<sub>type</sub>>::<action> e.g. &ldquo;mark::createMark&rdquo;
3.  the unset states for variables that we store in the socket should be nil in the fallthrough cases for the functions. This is so that we don&rsquo;t need to explicitly check if that atom is in the socket assigned state prior to reading it.

<a id="orgeb93e45"></a>

## Refactor Tasks

<a id="org5220093"></a>

### [X] the reflector mutator doesn&rsquo;t seem to be working

The learning here is that if there&rsquo;s no diff in the changeset then nothing gets written to the db, so you can&rsquo;t just force a db write like that.
 see here
<file:///Users/rtshkmr/Projects/vyasa/lib/vyasa_web/components/contexts/read.ex>

1.  I think the &ldquo;mutate&rdquo; word is wrong. It should be &ldquo;commit<sub>draft</sub><sub>sheaf</sub>&rdquo; / &ldquo;commit<sub>sheaf</sub><sub>changes</sub>&rdquo;

    -   it only deals with a draft sheaf, no publishing
    -   it&rsquo;s actually just committing changes in
    -   we can put the ui and reflector state changes in the same place

<a id="orgb196120"></a>

### [X] the nested form problem

-   CLOSING NOTE <span class="timestamp-wrapper"><span class="timestamp">[2024-10-25 Fri 17:30] </span></span>   
    This is resolved, refer to [this commit](https://github.com/ve1ld/vyasa/pull/126/commits/5680f2017942f999b6441d9005689e2563ca276d)

I&rsquo;m currently just shimming a form components so that i can keep forms as &ldquo;stateless&rdquo; and within function components.

Here&rsquo;s an example of this shim

    def mark_display(assigns) do
      ~H"""
      
      <div class="border-l border-brand-light pl-2">
        <%= if @mark.state == :live do %>
          <.form
            for={%{}}
            id={"mark-display-form-" <> @id}
            phx-hook="PreventEventPropagation"
            data-selector={"#mark-display-form-" <> @id}
            data-event-name="submit"
            phx-submit="mark::editMarkContent"
            phx-value-mark_id={@mark.id}
            phx-target={@marks_target}
          >
            <div
              id={"mark-container-" <>
            @mark.id <> "-" <> @id}
              class="mb-2 bg-brand-light rounded-lg shadow-sm p-1 border-l-2 border-brand flex justify-between items-start"
            >
              <div
                :if={@is_editable?}
                id={"ordering-button-group-"<> @mark.id <> "-" <> @id}
                class="flex flex-col items-center"
              >
                <button
                  phx-click="dummy_event"
                  phx-target={@marks_target}
                  class="p-1 hover:bg-gray-200 rounded"
                  aria-label="Up Arrow"
                  type="button"
                >
                  <.icon
                    name="custom-icon-sort-up"
                    class="w-5 h-5 text-brand-dark hover:bg-brand rounded-full p-1"
                  />
                </button>
                
                <div class="mx-1 text-center text-md font-light"><%= @mark.order %></div>
                <button
                  phx-click="dummy_event"
                  type="button"
                  phx-target={@marks_target}
                  class="p-1 hover:bg-gray-200 rounded"
                  aria-label="Down Arrow"
                >
                  <.icon
                    name="custom-icon-sort-down"
                    class="w-5 h-5 text-brand-dark hover:bg-brand rounded-full p-1"
                  />
                </button>
              </div>
              <div
                id={"mark-content-container-" <> @mark.id <> "-" <> @id}
                class="h-full w-full flex-grow mx-2 pt-2"
              >
                <%= if !is_nil(@mark) && !is_nil(@mark.binding) && !is_nil(@mark.binding.window) && @mark.binding.window.quote !== "" do %>
                  <span class="block mb-1 text-sm italic text-secondary">
                    "<%= @mark.binding.window.quote %>"
                  </span>
                <% end %>
                <%= if is_binary(@mark.body) do %>
                  <div class="flex-grow h-full">
                    <.mark_body
                      id={@mark.id <> "-" <> @id}
                      mark_ui={@mark_ui}
                      body_content={@mark.body}
                    />
                  </div>
                <% end %>
              </div>
              <div
                :if={@is_editable?}
                id={"mark-edit-actions-button-group-" <> @mark.id <> "-" <> @id}
                class="h-full flex flex-col ml-2 space-y-2 justify-between"
              >
                <button
                  phx-click="mark::tombMark"
                  type="button"
                  phx-target={@marks_target}
                  phx-value-id={@mark.id}
                  title="Delete"
                  class="p-3 hover:bg-gray-200 rounded flex items-center justify-center"
                  aria-label="Delete"
                >
                  <.icon name="hero-x-mark" class="w-5 h-5 text-brand-dark font-bold" />
                </button>
                <%= if not @mark_ui.is_editing_content? do %>
                  <button
                    phx-click="ui::toggle_is_editing_mark_content?"
                    type="button"
                    phx-target={@marks_target}
                    phx-value-mark_id={@mark.id}
                    class="p-3 hover:bg-gray-200 rounded flex items-center justify-center"
                    aria-label="Toggle edit mark body"
                  >
                    <.icon name="custom-icon-recent-changes-ltr" class="w-5 h-5 text-brand-dark" />
                  </button>
                <% else %>
                  
                  <button
                    type="submit"
                    class="p-3 hover:bg-gray-200 rounded flex items-center justify-center"
                    aria-label="Edit mark body"
                  >
                    <.icon name="hero-bookmark" class="w-5 h-5 text-brand-dark" />
                  </button>
                <% end %>
              </div>
            </div>
          </.form>
        <% end %>
      </div>
      """
    end

The problem with that is, when i render a component B that gets rendered within component A and both are forms, then there&rsquo;s a case of
nested forms.

Now, based on the deafult dom behaviour, any event is typically bubbled up the branch of the tree. This means that if both A and B are forms, then
a submit event for B will bubble up to A and hence A&rsquo;s form will also interpret a submit event.

This is a problem because a form is submitted without the user intending to do so.

A hacky work-around would be to simply use JS hooks and prevent this propagation (bubbling) and this seems to work.
However, we desire a longer term solution.

Before looking at solutions,
some questions:

1.  are nested forms an antipattern and if so, why?
2.  does it make sense to use this JS hook to prevent the propagation of the event to avoid this problem described above?

1.  Solutions

    tldr; of what a good solution would look like:
    
    -   submittable function components that can make edits is what we want
    
    -   we also want function components that can render each other (nested doms)
    
    -   we already have the form html dom constraint, so we won&rsquo;t be doing any hacks for that (thought it&rsquo;s possible)
    
    -   we also need something long-term of a solution so we don&rsquo;t want to add any unnessary constraints.
    
    1.  A: make the forms flat and not nested.
    
        We already know that this works, but let&rsquo;s go with B first
    
    2.  B: don&rsquo;t make it a form [attempting now]
    
        This means that you&rsquo;d have to use JS hooks to capture the value within the textarea and then use that value to construct the payload for the event that the submission would have been for.
        
        1.  Conclusions on this
        
            1.  this works great. It can be nested within another function component that contains a form if we require, for whatever reason and it will still work
            2.  This doesn&rsquo;t add a ui constraint for us that &ldquo;all function components that contain forms must always be siblings&rdquo;, so we don&rsquo;t need to do css wizardry and get the UI working for  those instances.

<a id="org1f2640d"></a>

### [X] Granular UI State management

-   CLOSING NOTE <span class="timestamp-wrapper"><span class="timestamp">[2024-10-25 Fri 22:30] </span></span>   
    Nicely done!!

We now have a nice commit<sub>marks</sub><sub>in</sub><sub>reflector</sub>.
We just want to effect the DB writes here so that shall be its main responsiblity.

Potentially, this also happens to be a good place to update the ui, since the draft reflector will be set here. However, doing so actually fattens the function up because it does ui state changes as well.

To determine if the roles of this function should be changed, we consider the current callign contexts for the function:

1.  when doing an edit mark content ([ref](file:///Users/rtshkmr/Projects/vyasa/lib/vyasa_web/components/contexts/read.ex))
    -   DONE ui should do a ui<sub>toggle</sub><sub>is</sub><sub>editing</sub><sub>mark</sub><sub>content</sub>()
2.  when creating a mark ([ref](file:///Users/rtshkmr/Projects/vyasa/lib/vyasa_web/components/contexts/read.ex))
    -   ui should insert this new entry to the ui map that&rsquo;s in the marks ui state
        ui<sub>add</sub><sub>mark</sub>()
3.  when creating a mark on the same binding ([ref](file:///Users/rtshkmr/Projects/vyasa/lib/vyasa_web/components/contexts/read.ex))
    -   same as 2.
4.  when tombing a mark ([ref](file:///Users/rtshkmr/Projects/vyasa/lib/vyasa_web/components/contexts/read.ex))
    -   ui<sub>remove</sub><sub>mark</sub>()

The ui change should be happening outside of this commit function. It will be called from the handle<sub>events</sub>, we can expose wrapper functions within the read context and within the ui state to handle these.

Next,
The commit function also calls a maybe<sub>prepend</sub><sub>draft</sub><sub>reflector</sub> and this needs to be updated too. ([ref](file:///Users/rtshkmr/Projects/vyasa/lib/vyasa_web/components/contexts/read.ex)) it should also do an addition just like before.

The best way to do this is to just expose register/deregister functions at various levels of abstraction.

I think this causes a new bug in the ui state management
ref
<file:///Users/rtshkmr/Projects/vyasa/lib/vyasa_web/components/contexts/read.ex>

<a id="orgd6e6faa"></a>

### [X] ui function interface

-   CLOSING NOTE <span class="timestamp-wrapper"><span class="timestamp">[2024-10-25 Fri 20:49] </span></span>   
    the read context has ui<sub>\*</sub> functions that serve as wrappers

the ui update should have a simpler interface for it, so that we don&rsquo;t need to do this
<file:///Users/rtshkmr/Projects/vyasa/lib/vyasa_web/components/contexts/read.ex>

</details>

### CODE REVIEW

File: lib/vyasa_web/components/contexts/read.ex [hunk 1]
```diff
@@ -219,11 +176,16 @@ defmodule VyasaWeb.Context.Read do
     socket
   end
 
+  # fallthrough
   defp sync_media_session(socket) do
     socket
   end
 
   @doc """
+  TODO: this functionis still a WIP, will be looked at when we are merging w the permalinking piece
```
Comment (rtshkmr) [inline lib/vyasa_web/components/contexts/read.ex:185]: this todo will be there until we get to merging discussions with permalinking

File: lib/vyasa_web/components/contexts/read.ex [hunk 2]
```diff
@@ -278,10 +238,11 @@ defmodule VyasaWeb.Context.Read do
...
     socket
-    # |> assign(reply_to: nil)
-    |> assign(
-      reply_to: hd(sangh_id |> Vyasa.Sangh.get_sheafs_by_session(%{traits: ["published"]}))
-    )
+    |> assign(reply_to: nil)
+
+    # |> assign(
+    #   reply_to: hd(sangh_id |> Vyasa.Sangh.get_sheafs_by_session(%{traits: ["published"]}))
+    # )
```
Comment (rtshkmr) [inline lib/vyasa_web/components/contexts/read.ex:245]: therefore, these are just for testing

File: lib/vyasa_web/components/contexts/read.ex [hunk 3]
```diff
@@ -888,108 +995,143 @@ defmodule VyasaWeb.Context.Read do
...
+    } =
+      Map.values(verses)
+      |> Enum.find(fn v -> not is_nil(v.binding) end)
+
     socket
     |> stream_insert(
       :verses,
       %{verses[v_id] | binding: binding}
     )
   end
```
Comment (rtshkmr) [inline lib/vyasa_web/components/contexts/read.ex:1051]: I've changed how this cascade works, to be frank I forgot why the older version would not work with my changes in this PR 🤔 

this new version should be non-breaking anyway

File: lib/vyasa_web/components/contexts/ui_state.ex [hunk 4]
```diff
@@ -144,25 +146,151 @@ defmodule VyasaWeb.Context.Components.UiState.Marks do
...
-
-  # import Phoenix.Component, only: [assign: 2]
+defmodule VyasaWeb.Context.Components.UiState.Sheaf do
+  alias Vyasa.Sangh.{Sheaf, Mark}
+  alias VyasaWeb.Context.Components.UiState.Marks, as: MarksUiState
+  alias VyasaWeb.Context.Components.UiState.Sheaf, as: SheafUiState
 
+  # QQ: @ks0m1c i haven't thought of whether the is_active should be duplicated from
+  # the actual ui state and kept within the SheafUiState or not.
+  # @rtshkmr to decide later, and remove this flag  if unnecessary
```
Comment (rtshkmr) [inline lib/vyasa_web/components/contexts/ui_state.ex:156]: i duplicated it.

I want to de-couple the active state in the sheaf from this ui active state so that @ init phase, we will have at least one active sheaf (because we use the active value for it). however, we might want more than one "active" (from the pov of ui) sheafs to be displayed.

File: assets/js/hooks/prevent_dom_propagation.js
[the remark this answers is not in the document]
Comment (ks0m1c) [reply assets/js/hooks/prevent_dom_propagation.js:1]: purge it for now we can recover if needed

**Review by ks0m1c — COMMENTED** [review]
LGTM!

File: lib/utils/stream.ex [hunk 5]
```diff
@@ -0,0 +1,41 @@
+defmodule Utils.Stream do
```
[the remark this answers is not in the document]
Comment (rtshkmr) [reply lib/utils/stream.ex:1]: makes sense, will make this change now

File: lib/vyasa_web/components/contexts/ui_state.ex (hunk 4 above)
Comment (rtshkmr) [reply lib/vyasa_web/components/contexts/ui_state.ex:156]: just realised that because state struct is doing tracking, this can be incremental for the current browser sessoin, so we can make it always additive. 

Such patterns will be useful for doing a universal contextual back-button throughout the platform, similar to what redux does in the react world where it's atomic ui changes that can be reloaded / replayed

**Comment by rtshkmr** [comment]
> LGTM!

give explicit approval m8

File: lib/utils/stream.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (rtshkmr) [reply lib/utils/stream.ex]: done, marking as resolved

**Review by ks0m1c — APPROVED** [review]
